agent/xds/routes.go

```diff
@@ -2,18 +2,301 @@ package xds
+	}, nil
+}
+
+func makeRouteMatchForDiscoveryRoute(discoveryRoute *structs.DiscoveryRoute, protocol string) envoyroute.RouteMatch {
+	switch protocol {
+	case "http", "http2":
+		// The only match stanza is HTTP.
+	default:
+		return makeDefaultRouteMatch()
+	}
+
+	match := discoveryRoute.Definition.Match
+	if match == nil || match.IsEmpty() {
+		return makeDefaultRouteMatch()
+	}
+
+	em := envoyroute.RouteMatch{}
+
+	switch {
+	case match.HTTP.PathExact != "":
+		em.PathSpecifier = &envoyroute.RouteMatch_Path{Path: "/"}
+	case match.HTTP.PathPrefix != "":
+		em.PathSpecifier = &envoyroute.RouteMatch_Prefix{Prefix: "/"}
+	case match.HTTP.PathRegex != "":
+		em.PathSpecifier = &envoyroute.RouteMatch_Regex{Regex: "/"}
+	default:
+		em.PathSpecifier = &envoyroute.RouteMatch_Prefix{Prefix: "/"}
+	}
+
+	if len(match.HTTP.Header) > 0 {
+		em.Headers = make([]*envoyroute.HeaderMatcher, 0, len(match.HTTP.Header))
+		for _, hdr := range match.HTTP.Header {
+			eh := &envoyroute.HeaderMatcher{
+				Name: hdr.Name,
+			}
+
+			switch {
+			case hdr.Exact != "":
+				eh.HeaderMatchSpecifier = &envoyroute.HeaderMatcher_ExactMatch{
+					ExactMatch: hdr.Exact,
+				}
+			case hdr.Regex != "":
+				eh.HeaderMatchSpecifier = &envoyroute.HeaderMatcher_RegexMatch{
+					RegexMatch: hdr.Regex,
+				}
+			case hdr.Prefix != "":
+				eh.HeaderMatchSpecifier = &envoyroute.HeaderMatcher_PrefixMatch{
+					PrefixMatch: hdr.Prefix,
+				}
+			case hdr.Suffix != "":
+				eh.HeaderMatchSpecifier = &envoyroute.HeaderMatcher_SuffixMatch{
+					SuffixMatch: hdr.Suffix,
+				}
+			case hdr.Present:
+				eh.HeaderMatchSpecifier = &envoyroute.HeaderMatcher_PresentMatch{
+					PresentMatch: true,
+				}
+			case hdr.Invert: // THIS HAS TO BE LAST
+				eh.HeaderMatchSpecifier = &envoyroute.HeaderMatcher_PresentMatch{
+					// We set this to the misleading value of 'true' here
+					// because we'll generically invert it next.
+					PresentMatch: true,
+				}
+			default:
+				continue // skip this impossible situation
+			}
+
+			if hdr.Invert {
+				eh.InvertMatch = true
+			}
+
+			em.Headers = append(em.Headers, eh)
+		}
+	}
+
+	if len(match.HTTP.QueryParam) > 0 {
+		em.QueryParameters = make([]*envoyroute.QueryParameterMatcher, 0, len(match.HTTP.QueryParam))
+		for _, qm := range match.HTTP.QueryParam {
+			eq := &envoyroute.QueryParameterMatcher{
+				Name:  qm.Name,
+				Value: qm.Value,
+				Regex: makeBoolValue(qm.Regex),
+			}
+
+			em.QueryParameters = append(em.QueryParameters, eq)
+		}
+	}
+
+	return em
+}
+
+func makeDefaultRouteMatch() envoyroute.RouteMatch {
+	return envoyroute.RouteMatch{
+		PathSpecifier: &envoyroute.RouteMatch_Prefix{
+			Prefix: "/",
+		},
+		// TODO(banks) Envoy supports matching only valid GRPC
+		// requests which might be nice to add here for gRPC services
+		// but it's not supported in our current envoy SDK version
+		// although docs say it was supported by 1.8.0. Going to defer
+		// that until we've updated the deps.
```

Comment:
This might be doable now we updated for 1.10.0. No action needed just noting it.